securetar/__init__.py

```diff
@@ -131,26 +143,66 @@ def __enter__(self) -> tarfile.TarFile:
         self._encrypt = self._aes.encryptor()
         self._padder = padding.PKCS7(BLOCK_SIZE_BITS).padder()
 
-        self._tar = tarfile.open(
-            fileobj=self,
-            mode=self._tar_mode,
-            dereference=False,
-            bufsize=self._bufsize,
-        )
-        return self._tar
-
     def __exit__(self, exc_type, exc_value, traceback) -> None:
         """Close file."""
         if self._tar:
             self._tar.close()
             self._tar = None
+        self._close_file()
+
+    def _close_file(self) -> None:
+        """Close file."""
         if self._file:
             if not self._mode.startswith("r"):
                 self._file.write(self._encrypt.update(self._padder.finalize()))
             if not self._fileobj:
                 self._file.close()
             self._file = None
 
+    @contextmanager
+    def decrypt(self, tarinfo: tarfile.TarInfo) -> Generator[BinaryIO, None, None]:
+        """Decrypt inner tar.
+
+        This is a helper to decrypt data and discard the padding.
+        """
+
+        class DecryptInnerTar:
+            """Decrypt inner tar file."""
+
+            def __init__(self, parent: SecureTarFile) -> None:
+                """Initialize."""
+                self._parent = parent
+                self._pos = 0
+                self._size = tarinfo.size
+                self._tail = b""
+
+            def read(self, size: int = 0) -> bytes:
+                """Read data."""
+                if self._tail:
+                    # Finish reading tail
+                    data = self._tail[:size]
+                    self._tail = self._tail[size:]
+                    return data
+
+                data = self._parent.read(size)
+                self._pos += len(data)
+                if not data or self._size - self._pos > 16:
+                    return data
+
+                # Last block, read tail and discard padding
+                data += self._parent.read(self._size - self._pos)
+                padding_len = data[-1]
+                data = data[:-padding_len]
+                self._tail = data[size:]
```

Comment:
We could check the padding is as expected and raise `ValueError` if invalid.
Maybe we should add error handling in a follow-up?